pkg/decoder/decoder.go

```diff
@@ -7,102 +7,205 @@ package decoder
+const countDownDuration = 1 * time.Second
 
-// NewPayload returns a new decoder payload
-func NewPayload(content []byte, offset int64) *Payload {
-	return &Payload{content, offset}
-}
+// maxMessageLen represents the maximum length for a message
+var maxMessageLen = config.MaxMessageLen
 
-// Decoder splits raw data based on `\n`, and sends those messages to a channel
+// Decoder splits raw data into messages using '\n' for single-line logs and lineRe for multi-line logs
+// and sends them to a channel
 type Decoder struct {
 	InputChan  chan *Payload
 	OutputChan chan message.Message
+
 	msgBuffer  *bytes.Buffer
+	lineBuffer *bytes.Buffer
+
+	timer  *time.Timer
+	lineRe *regexp.Regexp
 }
 
 // InitializeDecoder returns a properly initialized Decoder
-func InitializedDecoder() *Decoder {
+func InitializeDecoder(source *config.IntegrationConfigLogSource) *Decoder {
+	var lineRe *regexp.Regexp
+	for _, rule := range source.ProcessingRules {
+		switch rule.Type {
+		case config.MULTILINE:
+			lineRe = rule.Reg
+		}
+	}
+
 	inputChan := make(chan *Payload)
 	outputChan := make(chan message.Message)
-	return New(inputChan, outputChan)
+	return New(inputChan, outputChan, lineRe)
 }
 
 // New returns an initialized Decoder
-func New(InputChan chan *Payload, OutputChan chan message.Message) *Decoder {
-	var msgBuf bytes.Buffer
+func New(InputChan chan *Payload, OutputChan chan message.Message, lineRe *regexp.Regexp) *Decoder {
+	var msgBuffer, lineBuffer bytes.Buffer
 	return &Decoder{
 		InputChan:  InputChan,
 		OutputChan: OutputChan,
-		msgBuffer:  &msgBuf,
+		msgBuffer:  &msgBuffer,
+		lineBuffer: &lineBuffer,
+		lineRe:     lineRe,
 	}
 }
 
+// isMultiLineEnabled returns true if the decoder supports multi-line logs
+func (d *Decoder) isMultiLineEnabled() bool {
+	return d.lineRe != nil
+}
+
 // Start starts the Decoder
 func (d *Decoder) Start() {
 	go d.run()
 }
 
+// Stop stops the Decoder
+func (d *Decoder) Stop() {
+	close(d.InputChan)
+}
+
 // run lets the Decoder handle data coming from the InputChan
 func (d *Decoder) run() {
 	for data := range d.InputChan {
-		d.decodeIncomingData(data.content, data.offset)
+		if d.isMultiLineEnabled() {
+			d.ackIncomingData()
+			endedWithEOL := d.decodeIncomingData(data.content, data.context)
+			if endedWithEOL {
+				d.ackEndIncomingData(data.context)
+			}
+		} else {
+			d.decodeIncomingData(data.content, data.context)
+		}
+
 	}
 	d.OutputChan <- message.NewStopMessage()
 }
 
-// Stop stops the Decoder
-func (d *Decoder) Stop() {
-	close(d.InputChan)
+// ackIncomingData stops the timer which flushes the multi-line buffer as we got new data
+func (d *Decoder) ackIncomingData() {
+	if d.timer != nil {
+		d.timer.Stop()
+	}
+}
+
+// ackEndIncomingData starts the timer which flushes the multi-line buffer
+func (d *Decoder) ackEndIncomingData(context PayloadContext) {
+	d.timer = time.AfterFunc(countDownDuration, func() {
```

Comment:
I'd be interested in knowing if that's going to be costly (scheduling/stopping timers that often)? Another option is to have a global tick (every 500ms for example) that flushes all pending line which were started more than 1s ago.

Reply:
+1